New receipt and more flexible receipt reading

valentk777 opened the pull request:
Hello, this is my pull request.
* Better and almost universal receipt text reading was implemented
* Date for receipt was added
* data for testing was added (also named as yes_* - means is working now and no_*  - some issue has been founded)
* structure of receipt was changed
* cosmetic changes in code has been made
* data about shops to local DB was added
* shops reading/getting from DB was implemented

Commit:
testData added

Commit:
NewReceipt Format style created;
Folder for ImageFormating created
First stepts to more flexible receipt data reading

Commit:
fix 4 unit test and stop other to work until new receipt data reading will be impplemented

Commit:
more flexible receipt reading

Commit:
Fixed problems with reading

Commit:
Fixed time showing, fixed data

Commit:
data retested
get date from almost every receipt - done
mirror changes for better receipt reading

valentk777 commented:
Later I will add unit test for my code part. Now it is commended. sorry for that. and more interesting thing is our image formatting part. it is not working as expected. I do not know why pictures is not rotated as we expected or cropped, but now I made it manually and I have a original data for future testing. also, some receipt was read as mess because a letter was too big or too small. Need to think about this issue in the future :)

Commit:
Working with data

MariusKurbakovas commented on GREEDY.UnitTests/DataManagersUnitTests/UnitTestDataConverter.cs:
```diff
@@ -1,82 +1,80 @@
+//            //Assert.Equal((decimal)0.60, dataConverter.ReceiptToItemList(receipt)[1].Price);
+//        }
+//    }
+//}
```
You told me yourself not to leave commented code

valentk777 replied:
Sure. I create this pull request just to get your oppinion about main changes. Unit test will be added later

MariusKurbakovas commented on GREEDY.UnitTests/DataManagersUnitTests/UnitTestReceiptCreating.cs:
```diff
@@ -0,0 +1,57 @@
+//            Assert.Equal(data, shopName);
+//        }
+//    }
+//}
```
lots and lots of commented code

MariusKurbakovas commented on GREEDY/Data/ShopDataModel.cs:
```diff
@@ -10,6 +10,7 @@ public class ShopDataModel
         [Required]
         public string Name { get; set; }
         public string Location { get; set; }
+        public string SubName { get; set; }
```
Can you explain what is SubName?

valentk777 replied:
I add comment about subname in shop method. Subname is like a name how that shop is named in formal way (like UAB MAXIMA, UAB PALINK = iki) and other. Sometimes name as we know not the same as in formal documents

MariusKurbakovas commented on GREEDY/DataManagers/DataConverter.cs:
```diff
@@ -1,122 +1,127 @@
-                dr[1] = item.Price;
-                dr[2] = item.Category;
-                dataTable.Rows.Add(dr);
+                Console.WriteLine("Nepavyko nuskaityti cekio");
```
This shouldn't be a console write line. If no item was created there are two options one is to return null other is throw an exception. Then the controller that is calling this method should deal with it (I believe returning null would be better because exceptions are for errors and it is not really an error that no items have been read)

valentk777 replied:
We has talked about it. We need to create information dialog for a user. This is just for testing. Can you create an informative error windiw for user? :)

MariusKurbakovas commented on GREEDY/DataManagers/DataConverter.cs:
```diff
@@ -1,122 +1,127 @@
+            //        dataTable.Rows.Add(dr);
+            //    }
+            //    return dataTable;
+            //}
```
Commented code (will we still really need this). If yes maybe we can store it somewhere on github without merging it to master. (for example a copy of this branch)

MariusKurbakovas commented on GREEDY/DataManagers/ReceiptCreating.cs:
```diff
@@ -0,0 +1,105 @@
+            }
+        }
+
+        public static List<Shop> GetExistingShop()
```
We are puting the methods that are writing/reading from database in separate classes, with their name ending Manager. See other code for an example.

valentk777 replied:
Agree. Thanks for comment. This part was a question for me (to create it there or in new file).

MariusKurbakovas commented on GREEDY/ImagePreparation/ImageFormating.cs:
```diff
@@ -42,10 +42,16 @@ private Bitmap Crop(Bitmap bitmap, double[] cornerPoints)
+                if (list.Any())
+                    return list.MaxBy(x => x.Height * x.Width); // this guy throws an exception if list is null.
+            }
+            catch (Exception e)
```
Why is this catch called? There can be no exception in the try clause

MariusKurbakovas commented on GREEDY/ImagePreparation/ImageFormating.cs:
```diff
@@ -42,10 +42,16 @@ private Bitmap Crop(Bitmap bitmap, double[] cornerPoints)
+            try
+            {
+                if (list.Any())
+                    return list.MaxBy(x => x.Height * x.Width); // this guy throws an exception if list is null.
```
Just like to point out that we are using the microsoft coding standard which i believe requires { } clauses and new lines even when there is one line in the block (easier to read that way)

valentk777 replied:
Can you show me what is wrong there? 😅 sorry, but I cannot understand. I just added try and catch. No other changes was made

MariusKurbakovas reviewed (commented):
Didn't test how it works, but look pretty good

AidasKil commented on GREEDY/DataManagers/ReceiptCreating.cs:
```diff
@@ -0,0 +1,105 @@
+                    {'Name': 'MAXIMA','Adress': '', 'Subname': ''},
+                    {'Name': 'RIMI','Adress': '','Subname': ''},
+                    {'Name': 'LIDL','Adress': '','Subname': ''}
+                ]");
```
Why would you hard code it as json only to convert to object? If it is hardcoded you can just create it as object

Commit:
Setup migrations
After making changes to database run entityframework\add-migration [Name of the migration] in nuget console.

To update your database run entityframework\update-database in nuget console

To add default data to database add it in Configuration.Seed method(just make sure it doesn't duplicate data as it is called every time database is updated)

valentk777 commented:
Aidas, thanks for your changes!

Commit:
cosmetic changes has been made special for pull request comment. Unit test will be added after Aidas changes, because I need to use DatabaseMock.cs
